File: test/regression/roundabout_connections.js
'use strict';
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');
const repo = path.resolve(__dirname, '../..');
const data = JSON.parse(fs.readFileSync(process.argv[2]));
const sandbox = {window: {TM: {}, location: {pathname: '/en/map'}}, console};
vm.createContext(sandbox);
for (const file of ['converter/road-names.js', 'web/src/scripts/map-desc-ways.js']) {
  vm.runInContext(fs.readFileSync(path.join(repo, file), 'utf8'), sandbox);
}
for (const locale of ['en', 'fi', 'de', 'es', 'nl']) {
  const translations = JSON.parse(fs.readFileSync(path.join(repo, 'web/locales', locale, 'tm.json')));
  const render = input => sandbox.window.TM.mapDescWays.buildModel(input,
    {t: (key, fallback) => translations[key] || fallback}, {section: 'roads'});
  const item = (items, name) => items.find(entry => entry.summaryTitle.includes(name));
  const lines = entry => Array.from(entry.lines, line => line.parts.map(part => part.text).join(''));
  const result = render(data);
  const first = lines(item(result, 'First Road'));
  const second = lines(item(result, 'Second Road'));
  const roundaboutText = translations.map_content_connects_to_roundabout;
  assert.strictEqual(first.filter(line => line === roundaboutText).length, 1);
  assert.strictEqual(second.filter(line => line === roundaboutText).length, 1);
  assert(!first.some(line => line.includes('Side Street')), 'No invented connection across roundabout');
  assert(second.includes(translations.map_content_connects_to_way.replace('__way__', 'Side Street')));
  assert(!first.some(line => line === translations.map_content_connects_to_type_many
    .replace('__count__', '2').replace('__type__', translations.map_content_way_type_plural_A1_secondary_roads)),
    'Roundabout segments must not also be counted as roads');
  assert(lines(item(result, 'Bridge Road')).includes(
    translations.map_content_connects_to_roundabouts.replace('__count__', '2')),
    'Distinct roundabouts joined by a road are counted separately');
  assert.deepStrictEqual(Array.from(item(result, 'First Road').attrs.filterRefs), ['way:1']);

  const named = JSON.parse(JSON.stringify(data));
  for (const group of named.A.subclasses[0].groups) {
    for (const bucket of group.visibleGeometry) for (const segment of bucket.segments) {
      for (const event of segment.events) for (const contact of event.connections || []) {
        if (contact.roundabout && contact.roundabout.id === 'roundabout:way:10') {
          contact.roundabout.name = 'Central Circle';
        }
      }
    }
  }
  const namedFirst = lines(item(render(named), 'First Road'));
  assert(namedFirst.includes(translations.map_content_connects_to_named_roundabout.replace('__name__', 'Central Circle')));
  assert(!namedFirst.includes(roundaboutText));
}
console.log('Split roundabout identity and localized connection descriptions passed');

File: test/regression/roundabout_connections.py
#!/usr/bin/env python3
"""Check converter roundabout identity and its actual localized UI description."""
import json
from pathlib import Path
import subprocess
import sys

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))
from converter.map_desc.map_desc_render import (
    _build_connections_index, _build_connectors_by_coord_key,
    _collect_inferred_named_connectors, _build_way_groups)


def way(osm_id, coords, name=None, roundabout=False):
    tags = {'highway': 'secondary'}
    if name:
        tags['name'] = name
    if roundabout:
        tags['junction'] = 'roundabout'
    return {'osmId': osm_id, 'osmType': 'way', 'tags': tags,
            '_classification': {'subClass': 'A1_secondary_roads'},
            'geometry': {'type': 'line_string', 'coordinates': coords},
            'visibleGeometry': [coords]}


def main():
    # One ring split into three ways, two approaches, and a separate named street
    # joining only the second approach. Connector nodes deliberately absent.
    ring = [way(10, [[0, 0], [10, 0]], roundabout=True),
            way(11, [[10, 0], [5, 10]], roundabout=True),
            way(12, [[5, 10], [0, 0]], roundabout=True)]
    roads = [way(1, [[-10, 0], [0, 0]], 'First Road'),
             way(2, [[20, 0], [10, 0]], 'Second Road'),
             way(3, [[20, 0], [20, -10]], 'Side Street')]
    # Another ring connected by an ordinary road must keep a separate identity.
    other = way(20, [[30, 0], [40, 0], [35, 10], [30, 0]], roundabout=True)
    bridge = way(4, [[5, 10], [30, 0]], 'Bridge Road')
    items = ring + roads + [other, bridge]
    grouped = {'A': {'roads': items}}
    index = _build_connections_index(grouped)
    records = {feature['osmId']: feature for values in index.values() for feature in values}
    identity = records[10]['roundabout']
    assert all(records[i]['roundabout'] == identity for i in [10, 11, 12])
    assert identity['name'] is None
    assert records[20]['roundabout']['id'] != identity['id']
    assert all('roundabout' not in records[i] for i in [1, 2, 3, 4])
    reordered = _build_connections_index({'A': {'roads': list(reversed(items))}})
    assert {f['osmId']: f for values in reordered.values() for f in values} == records

    # Clipping may separate the visible arcs; their source topology still
    # identifies one roundabout, without exposing clipped-away junctions.
    ring[0]['visibleGeometry'] = [[[1, 0], [9, 0]]]
    clipped = _build_connections_index(grouped)
    assert all(f['roundabout'] == identity for values in clipped.values()
               for f in values if f['osmId'] in [10, 11, 12])
    assert not any(f['osmId'] == 10 for f in clipped.get('0.000,0.000', []))
    ring[0]['visibleGeometry'] = [ring[0]['geometry']['coordinates']]

    # A genuine roundabout name can be carried on just one constituent way.
    ring[0]['tags']['name'] = 'Central Circle'
    named = _build_connections_index(grouped)
    assert all(f['roundabout']['name'] == 'Central Circle' for values in named.values()
               for f in values if f['osmId'] in [10, 11, 12])
    del ring[0]['tags']['name']

    connectors = _build_connectors_by_coord_key(_collect_inferred_named_connectors(index))
    groups = _build_way_groups(items, None, connectors, index, True)
    first = next(group for group in groups if group['label'] == 'First Road')
    contacts = [contact for bucket in first['visibleGeometry'] for segment in bucket['segments']
                for event in segment['events'] for contact in event.get('connections', [])]
    assert {contact['osmId'] for contact in contacts} == {1, 10, 12}
    assert sum(bool(contact.get('roundabout')) for contact in contacts) == 2
    output = Path(sys.argv[1]) / 'roundabout-connections.json'
    output.write_text(json.dumps({'A': {'subclasses': [{
        'key': 'A1_secondary_roads', 'kind': 'linear', 'groups': groups}]}}))
    subprocess.run(['node', str(REPO / 'test/regression/roundabout_connections.js'), str(output)], check=True)


if __name__ == '__main__':
    main()
